Feature/rebuild decision maker

s-azumi opened the pull request:
## Status
**DEVELOPMENT**

## Description
Please refer to autowarefoundation/autoware_ai#252 
I redesigned the state transition and divided it into three state machines Vehicle, Mission, Drive.
The details are as follows.
<img src="https://user-images.githubusercontent.com/38061530/49000214-33242880-f19c-11e8-8270-ac8e041c2942.jpg" width="100" title="Vehicle state"><img src="https://user-images.githubusercontent.com/38061530/49000212-33242880-f19c-11e8-9f82-4141fa674bb6.jpg" width="320" title="Mission state">
<img src="https://user-images.githubusercontent.com/38061530/49000211-33242880-f19c-11e8-9d21-722cdbd98cbc.jpg" width="480" title="Drive state">

## Related branch
`feature/last_wp_control` is needed to reach waypoint's end and transition to the MissionComplete state.
`feature/negative_velocity` is required for switchback.


## Todos
- [x] Tests
  - [x] amath_utils (#1757 To be merged)
  - [x] state_machine_lib (#1744 To be merged)
  - [x] decision_maker (#1958 To be merged)
  - [x] lane_planner (#2009 To be merged)
  - [x] waypoint_follower (#2028 To be merged)
- [x] Documentation https://github.com/CPFL/Autoware/blob/feature/rebuild_decision_maker/ros/src/computing/planning/decision/packages/decision_maker/README.md


## Steps to Test or Reproduce
NOTE: For state visualization, please add OverlayText plugin on rviz and select topics `decision_maker/operator_help_text`, `state/overlay_text`.

**Normal driving**
1. Launch Autoware and localization on real vehicle or wf_simulator
2. Launch `decision_maker` and lane_planner, astar_planner, waypoint_follower
3. When `VehicleReady` and `WaitOrder` state, launch `waypoint_loader`
4. On `DriveReady` state, push `Engage` button on DecisionMakerPannel
5. When the vehicle reaches the end of waypoint and stops, state Mission state transits to `WaitOrder` via the `MissionComplete`
6. You can repeat from 3. with other waypoint

**Lane change**
1. Start Normal driving with waypoint files necessary for lane change
2. On `CheckLeft/RightLane` state, push `Execute LaneChange` button on DecisionMakerPannel
3. The vehicle start lane change

NOTE: Safety check and automatically lane change functions are not implemented yet.

**Driving mission change**
1. Prepare waypoint files to change, and start Normal driving
2. On `Driving` state, push `Request mission change` button on DecisionMakerPannel
3. When the state becomes `DrivingMissionChange`, please load another waypoint for change
4. If it is possible to change the waypoint will switch, otherwise it will not be changed

NOTE: This function is quick implementation. Whether change is possible or not  depends on two parameters `dist_threshold`, `angle_threshold` between current_pose and nearest waypoint after change.

yk-fujii committed:
rebuild decision_maker

yk-fujii committed:
update maitainer for decision maker

yn-mrse reviewed (changes requested):
How nice! This feature worked steadily in my environment.
I can merge after the following fixes are completed.

messi49 committed:
fix emergency bug

s-azumi committed:
fixed state_machine_lib according to review.

s-azumi committed:
deleted unused function in amathutils.

s-azumi committed:
fixed decision_maker nodes according to review.

s-azumi committed:
Merge pull request #1699 from CPFL/feature/fix_twist_gate_emergency_bug

fix emergency bug

s-azumi committed:
fixed emergency bug of state transition and command sent from decision_maker_pannel.

moded dot file generation of state_machine.

Commit:
Fix error in Amathutils.cpp which caused error calculating distance between 3D points

yk-fujii committed:
Merge branch 'develop' of github.com:cpfl/autoware into feature/rebuild_decision_maker

yk-fujii committed:
Merge pull request #1703 from sgermanserrano/feature/rebuild_decision_maker

Fix error in Amathutils.cpp

yk-fujii committed:
Merge branch 'feature/rebuild_decision_maker' of github.com:cpfl/autoware into feature/rebuild_decision_maker

yk-fujii committed:
apply clang-format

s-azumi committed:
removed planner_selector and its related parameters

s-azumi committed:
add WaitReady state into Drive states, and fixed emergency bag.

s-azumi committed:
fixed TryAvoidance state bug.

s-azumi committed:
fixed stop button bug and mission change bug.

removed DTLANE from required vector_map.

s-azumi committed:
corrected the codes that where pointed out by sgermanserrano.

s-azumi committed:
removed unnecessary codes.

s-azumi committed:
removed unnecessary codes.

parameterized reset_count of stopline function.

kfunaoka reviewed (commented):
@s-azumi Would you review and merge https://github.com/CPFL/Autoware/pull/1651?

sgermanserrano committed:
Added unit tests for amathutils_lib (#1651)

* Added intial set of unit tests for amathutils_lib

s-azumi committed:
Merge branch 'develop' into feature/rebuild_decision_maker

s-azumi committed:
removed unnecessary codes.

add exception handling for getStateIDbyName function in state_context.cpp.

yk-fujii commented:
@s-azumi 
Could you please the following support?
- to remove clang-format commit for another packages
- to check code by cppcheck
- to create readme and overview documents
- to do enrich autowarefoundation/autoware_ai#252 

I think to clear for 1,2,3,4,5,8,9,10 comments.
**Also, we need to add unit test codes.**

@dejanpan  
> 7. How will you make sure that this does not break everything in Autoware since it is such a fundamental change?

This proposal(#1434) was submitted in August. 
But nobody has proposed any problems, any comments. 
So, we judged that there was no problem even if we replaced it.
In addition, the operation confirmation by the Integration test has been ongoing for a long time now.
Also, we can not continue to support current decision maker, because we have not used current decision_maker since six months ago and we will not use it from now on.

Commit:
Added tests for the new functions

Commit:
Removed extra space

Commit:
Updated to ROS Cpp Style guidelines

Commit:
Added license notice

mitsudome-r committed:
initial commit for decision_maker README.md

mitsudome-r committed:
modified README.md for decision_maker

kfunaoka committed:
Replace const double & with double in amathutils_lib

kfunaoka committed:
Merge branch 'develop' into feature/rebuild_decision_maker

kfunaoka commented:
At 7001f22, 84 files and the following components are changed:

- amathutils_lib -> difficult to split autowarefoundation/autoware_ai#479
- diagnostics_lib -> removed from this pull request
- state_machine_lib
- decision_maker
- lane_planner
- astar_planner -> removed from this pull request
- dp_planner -> removed from this pull request
- waypoint_follower
- waypoint_maker -> just change type and name -> split to autowarefoundation/autoware#1822 -> merged to develop.
- vector_map -> just make declaration public -> split to autowarefoundation/autoware#1821 -> merged to develop.
- autoware_config_msgs
- autoware_msgs
- vehicle_sender -> split to autowarefoundation/autoware#1820 -> reverted at 04b2a1b.
- autoware_rviz_plugins
- runtime_manager

```
$ git diff develop..feature/rebuild_decision_maker --name-only
ros/src/common/libs/amathutils_lib/CMakeLists.txt
ros/src/common/libs/amathutils_lib/include/amathutils_lib/amathutils.hpp
ros/src/common/libs/amathutils_lib/src/Amathutils.cpp
ros/src/common/libs/amathutils_lib/test/src/test_amathutils_lib.cpp
ros/src/common/libs/amathutils_lib/test/test_amathutils_lib.test
ros/src/common/libs/diagnostics_lib/diag_lib/include/diag_lib/diag_filter.h
ros/src/common/libs/diagnostics_lib/diag_lib/include/diag_lib/diag_info.h
ros/src/common/libs/diagnostics_lib/diag_lib/include/diag_lib/diag_macros.h
ros/src/common/libs/diagnostics_lib/diag_lib/include/diag_lib/diag_manager.h
ros/src/common/libs/diagnostics_lib/diag_lib/include/diag_lib/diag_manager_config.h
ros/src/common/libs/diagnostics_lib/diag_lib/include/diag_lib/diag_subscriber.h
ros/src/common/libs/diagnostics_lib/diag_lib/include/diag_lib/error_category.h
ros/src/common/libs/diagnostics_lib/diag_lib/include/diag_lib/rate_checker.h
ros/src/common/libs/diagnostics_lib/diag_lib/include/diag_lib/watchdog.h
ros/src/common/libs/diagnostics_lib/diag_lib/src/diag_filter.cpp
ros/src/common/libs/diagnostics_lib/diag_lib/src/diag_manager.cpp
ros/src/common/libs/diagnostics_lib/diag_lib/src/diag_subscriber.cpp
ros/src/common/libs/diagnostics_lib/diag_lib/src/rate_checker.cpp
ros/src/common/libs/diagnostics_lib/diag_lib/src/watchdog.cpp
ros/src/common/libs/diagnostics_lib/diag_lib/src/watchdog_node.cpp
ros/src/common/libs/diagnostics_lib/fake_autoware_nodes/include/fake_autoware_nodes/fake_publisher.h
ros/src/common/libs/diagnostics_lib/fake_autoware_nodes/include/fake_autoware_nodes/fake_subscriber.h
ros/src/common/libs/diagnostics_lib/fake_autoware_nodes/src/fake_publisher.cpp
ros/src/common/libs/diagnostics_lib/fake_autoware_nodes/src/fake_publisher_node.cpp
ros/src/common/libs/diagnostics_lib/fake_autoware_nodes/src/fake_subscriber.cpp
ros/src/common/libs/diagnostics_lib/fake_autoware_nodes/src/fake_subscriber_node.cpp
ros/src/common/libs/state_machine_lib/CMakeLists.txt
ros/src/common/libs/state_machine_lib/include/state_machine_lib/state.hpp
ros/src/common/libs/state_machine_lib/include/state_machine_lib/state_common.hpp
ros/src/common/libs/state_machine_lib/include/state_machine_lib/state_context.hpp
ros/src/common/libs/state_machine_lib/include/state_machine_lib/state_drive.hpp
ros/src/common/libs/state_machine_lib/include/state_machine_lib/state_emg.hpp
ros/src/common/libs/state_machine_lib/include/state_machine_lib/state_flags.hpp
ros/src/common/libs/state_machine_lib/include/state_machine_lib/state_main.hpp
ros/src/common/libs/state_machine_lib/src/state_context.cpp
ros/src/computing/planning/decision/packages/decision_maker/CMakeLists.txt
ros/src/computing/planning/decision/packages/decision_maker/README.md
ros/src/computing/planning/decision/packages/decision_maker/autoware_state_drive.yaml
ros/src/computing/planning/decision/packages/decision_maker/autoware_state_mission.yaml
ros/src/computing/planning/decision/packages/decision_maker/autoware_state_vehicle.yaml
ros/src/computing/planning/decision/packages/decision_maker/include/cross_road_area.hpp
ros/src/computing/planning/decision/packages/decision_maker/include/decision_maker_node.hpp
ros/src/computing/planning/decision/packages/decision_maker/include/decision_maker_param.hpp
ros/src/computing/planning/decision/packages/decision_maker/include/planner_selector.hpp
ros/src/computing/planning/decision/packages/decision_maker/interface.yaml
ros/src/computing/planning/decision/packages/decision_maker/launch/decision_maker.launch
ros/src/computing/planning/decision/packages/decision_maker/launch/planner_selector.launch
ros/src/computing/planning/decision/packages/decision_maker/nodes/decision_maker/cross_road_area.cpp
ros/src/computing/planning/decision/packages/decision_maker/nodes/decision_maker/decision_maker_node.cpp
ros/src/computing/planning/decision/packages/decision_maker/nodes/decision_maker/decision_maker_node_callback.cpp
ros/src/computing/planning/decision/packages/decision_maker/nodes/decision_maker/decision_maker_node_core.cpp
ros/src/computing/planning/decision/packages/decision_maker/nodes/decision_maker/decision_maker_node_decision.cpp
ros/src/computing/planning/decision/packages/decision_maker/nodes/decision_maker/decision_maker_node_init.cpp
ros/src/computing/planning/decision/packages/decision_maker/nodes/decision_maker/decision_maker_node_publish.cpp
ros/src/computing/planning/decision/packages/decision_maker/nodes/decision_maker/decision_maker_node_state_drive.cpp
ros/src/computing/planning/decision/packages/decision_maker/nodes/decision_maker/decision_maker_node_state_mission.cpp
ros/src/computing/planning/decision/packages/decision_maker/nodes/decision_maker/decision_maker_node_state_vehicle.cpp
ros/src/computing/planning/decision/packages/decision_maker/nodes/decision_maker/decision_maker_node_stateupdate.cpp
ros/src/computing/planning/decision/packages/decision_maker/nodes/planner_selector/planner_selector_node.cpp
ros/src/computing/planning/decision/packages/decision_maker/package.xml
ros/src/computing/planning/mission/packages/lane_planner/launch/lane_select.launch
ros/src/computing/planning/mission/packages/lane_planner/nodes/lane_rule/lane_rule.cpp
ros/src/computing/planning/mission/packages/lane_planner/nodes/lane_select/lane_select_core.cpp
ros/src/computing/planning/mission/packages/lane_planner/nodes/lane_select/lane_select_core.h
ros/src/computing/planning/motion/packages/astar_planner/launch/velocity_set.launch
ros/src/computing/planning/motion/packages/astar_planner/nodes/velocity_set/velocity_set.cpp
ros/src/computing/planning/motion/packages/dp_planner/include/dp_planner_core.h
ros/src/computing/planning/motion/packages/dp_planner/launch/dp_planner.launch
ros/src/computing/planning/motion/packages/dp_planner/nodes/dp_planner_core.cpp
ros/src/computing/planning/motion/packages/waypoint_follower/nodes/twist_gate/twist_gate.cpp
ros/src/computing/planning/motion/packages/waypoint_follower/nodes/wf_simulator/wf_simulator.cpp
ros/src/computing/planning/motion/packages/waypoint_maker/nodes/waypoint_loader/waypoint_loader_core.cpp
ros/src/data/packages/vector_map/include/vector_map/vector_map.h
ros/src/msgs/autoware_config_msgs/msg/ConfigDecisionMaker.msg
ros/src/msgs/autoware_msgs/CMakeLists.txt
ros/src/msgs/autoware_msgs/msg/LaneArray.msg
ros/src/msgs/autoware_msgs/msg/VehicleLocation.msg
ros/src/msgs/autoware_msgs/msg/WaypointState.msg
ros/src/socket/packages/vehicle_socket/nodes/vehicle_sender/vehicle_sender.cpp
ros/src/util/packages/autoware_rviz_plugins/CMakeLists.txt
ros/src/util/packages/autoware_rviz_plugins/plugin_description.xml
ros/src/util/packages/autoware_rviz_plugins/src/decision_maker_panel/decision_maker_panel.cpp
ros/src/util/packages/autoware_rviz_plugins/src/decision_maker_panel/decision_maker_panel.h
ros/src/util/packages/runtime_manager/scripts/computing.yaml
```

kfunaoka committed:
Revert vehicle_sender.cpp

kfunaoka committed:
Revert the copyright in wf_simulator.cpp

mitsudome-r committed:
updated README.md

mitsudome-r committed:
updated state description in readme

mitsudome-r committed:
fixed typo in state description in readme

mitsudome-r committed:
Update README.md

s-azumi committed:
change output dot file name of state_machine_lib.

s-azumi committed:
fixed function to get angle diff in drivingMissionCheck function.

s-azumi committed:
fix bug if the vehicle stops due to obstacle while StopLine state.

s-azumi committed:
remove unused state: Avoidance

s-azumi committed:
add state transition return from emergency to ready state

s-azumi committed:
add automatically engage and automatically mission change function, and those checkbox to runtime_manager

s-azumi committed:
Merge branch 'develop' into feature/rebuild_decision_maker

s-azumi committed:
remove unused variables & functions

kfunaoka commented:
@sgermanserrano @s-azumi @mitsudome-r I split some changes to autowarefoundation/autoware#1821 and autowarefoundation/autoware#1822. This pull request no longer has the changes of waypoint_maker and vector_map. I removed these packages from ToDo.

kfunaoka commented:
@s-azumi Would you revert some clang-format-only changes where you already made the revert commit👍

s-azumi committed:
reverted changes diagnostics_lib

kfunaoka reviewed (commented):
some changes are made

- lane_planner/nodes/lane_rule/lane_rule.cpp : the inline comment only
- lane_planner/nodes/lane_select/lane_select_core.cpp
- lane_planner/nodes/lane_select/lane_select_core.h

can be split or reverted

- dp_planner : all changes are enablePlannerDynamicSwitch (#1919) or white space changes.
- velocity_set : all changes are enablePlannerDynamicSwitch (#1919)
- lane_planner/nodes/lane_rule/lane_rule.cpp : most of lines except for the inline comment

kfunaoka committed:
Merge develop into feature/rebuild_decision_maker

kfunaoka committed:
Merge 'develop' into feature/rebuild_decision_maker

kfunaoka committed:
Revert dp_planner to split the changes into #1919

kfunaoka commented:
I've removed dp_planner from ToDo. The changes of dp_planner is reverted at 00679de and split into autowarefoundation/autoware#1919.

kfunaoka committed:
Revert dp_planner to split the changes into #1919

kfunaoka commented:
I've removed velocity_set from ToDo. The changes of velocity_set is reverted at 4e13070 and split into autowarefoundation/autoware#1919.

kfunaoka committed:
Rename directories of decision_maker_panel

kfunaoka committed:
Remove the target decision_maker_panel.cpp from state_panel

kfunaoka committed:
dd CMakeLists.txt and package.xml to decision_maker_panel

kfunaoka committed:
Remove decision_maker_panel in plugin_description from state_panel

kfunaoka committed:
Add plugin_description.xml to decision_maker_panel

s-azumi committed:
Insert <rviz plugin="${prefix}/plugin_description.xml"/> into package.xml

Co-Authored-By: kfunaoka <kfunaoka@gmail.com>

kfunaoka committed:
Add autoware_msgs to package.xml in decision_maker_panel and remove it from state_panel

kfunaoka commented:
autowarefoundation/autoware#1920 and autowarefoundation/autoware#1921 fix the CI errors as shown in autowarefoundation/autoware#1921.

kfunaoka committed:
[Fix] industrial_ci failure in feature/rebuild_decision_maker (#1921)

kfunaoka committed:
Revert removing enablePlannerDynamicSwitch in velocity_set.launch

kfunaoka committed:
Merge pull request #1920 from CPFL/feature/decision_maker_panel_package

Also merge pull request #1921

s-azumi committed:
change state name from PullOver to PullIn

change topic name from state/overlay_text to decision_maker/state_overlay
, from target_velocity_array to decision_maker/target_velocity_array

s-azumi committed:
refactor DecisionMakerPanel, and add buttons to the panel

s-azumi committed:
moded autoware_msgs/State for current decision_maker
add publisher for state_msg

kfunaoka committed:
Merge 'feature/rebuild_decision_maker' into feature/test_PR_1609

Commit:
Merge remote-tracking branch 'upstream/feature/test_PR_1609' into feature/test_amathutils

s-azumi committed:
fix transition bug of return command while emergency state

kfunaoka committed:
Merge pull request #1757 from sgermanserrano/feature/test_amathutils

kfunaoka committed:
Create twist_gate_node.cpp

s-azumi committed:
fix emergency bug of twist_gate

s-azumi committed:
fix typo in decision_maker_panel.cpp

s-azumi committed:
remove unused parameter

s-azumi committed:
update README.md and jpg files in docs folder

s-azumi committed:
chenge parameter name from use_management_system to use_fms

add parameters to initializer lists of DecisionMakerNode class

enabled parameter initialization from launch file

s-azumi committed:
add test of decision_maker

s-azumi committed:
fix bug in getEventStateFromWaypoint function

s-azumi committed:
apply clang-format

s-azumi committed:
Merge branch 'develop' into feature/rebuild_decision_maker

s-azumi committed:
Merge branch 'feature/rebuild_decision_maker' into feature/test_decision_maker

s-azumi committed:
update CMakeLists.txt and package.xml

Commit:
Added dependencies

Commit:
Added TestClass

s-azumi committed:
Merge pull request #1964 from sgermanserrano/fix/test_decision_maker

Fix/test decision maker

s-azumi committed:
add License notices

s-azumi committed:
add test of inLocalizationConvergence and fix bug of the function

s-azumi committed:
change the directory of files for testing

s-azumi committed:
add test of lane_select for the new function

s-azumi committed:
change directory of twist_gate.h

s-azumi committed:
changes for test

s-azumi committed:
add test for twist_gate

s-azumi committed:
Merge branch 'feature/rebuild_decision_maker' into feature/test_twist_gate

s-azumi committed:
apply clang-format

kfunaoka committed:
Merge pull request #2009 from CPFL/feature/test_lane_select

s-azumi committed:
Merge branch 'develop' into feature/rebuild_decision_maker

s-azumi committed:
add test for changes by rebuild_decision_maker

s-azumi committed:
appry clang-format

s-azumi committed:
add help text message

s-azumi committed:
fix thread in twist_gate.cpp

s-azumi committed:
Merge branch 'feature/rebuild_decision_maker' into feature/test_twist_gate

s-azumi committed:
revert changes of test_amathutils_lib.cpp

s-azumi committed:
update README.md

s-azumi committed:
Merge branch 'develop' into feature/rebuild_decision_maker

s-azumi committed:
Merge branch 'develop' into feature/rebuild_decision_maker

s-azumi committed:
Merge branch 'feature/rebuild_decision_maker' into feature/test_twist_gate

s-azumi committed:
Merge pull request #1958 from CPFL/feature/test_decision_maker

Feature/test decision maker

sgermanserrano commented:
@kfunaoka @gbiggs I agree with the approach after merging autowarefoundation/autoware#2028 
Nevertheless, I think knowledge of the underlying algorithm is needed for the development of tests so I'm not sure the depth of testing that a software engineering company can bring. IMO improved documentation of the algorithms/nodes would improve testing contributions from the wider community

s-azumi committed:
Merge branch 'develop' into feature/rebuild_decision_maker

s-azumi committed:
Merge branch 'feature/rebuild_decision_maker' into feature/test_twist_gate

s-azumi committed:
fix CMakeLists for colcon

s-azumi committed:
add notation "experimental"

fix bugs of state transitions:

* add mission check requirement

* add reset the flag "received_vased_lane_waypoint"

* change subscriber queue size

* add header.stamp into publishLampCmd

* add initialization of vector map in entryMissionCheck function

s-azumi committed:
add tests for existing functions

s-azumi committed:
change directry of includes files

yn-mrse committed:
fix parameter link of computing.yaml

s-azumi committed:
add tests

change directry of twist_gate.h

kfunaoka committed:
Merge pull request #2028 from CPFL/feature/test_twist_gate

kfunaoka committed:
Merge branch 'develop' into feature/rebuild_decision_maker
